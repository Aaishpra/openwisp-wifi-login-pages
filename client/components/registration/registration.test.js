/* eslint-disable prefer-promise-reject-errors */
import axios from "axios";
/* eslint-disable camelcase */
import {shallow} from "enzyme";
import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";

import getConfig from "../../utils/get-config";
import Registration from "./registration";

jest.mock("axios");

const defaultConfig = getConfig("default");
const createTestProps = props => {
  return {
    language: "en",
    orgSlug: "default",
    registration: defaultConfig.components.registration_form,
    privacyPolicy: defaultConfig.privacy_policy,
    termsAndConditions: defaultConfig.terms_and_conditions,
    authenticate: jest.fn(),
    match: {
      path: "default/registration",
    },
    ...props,
  };
};

describe("<Registration /> rendering", () => {
  let props;
  beforeEach(() => {
    props = createTestProps();
  });
  it("should render correctly", () => {
    props = createTestProps();
    const renderer = new ShallowRenderer();
    const component = renderer.render(<Registration {...props} />);
    expect(component).toMatchSnapshot();
  });
});

describe("<Registration /> interactions", () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<Registration {...props} />);
  });
  it("should change state values when handleChange function is invoked", () => {
    wrapper
      .find("#owisp-registration-username")
      .simulate("change", {target: {value: "test username", name: "username"}});
    expect(wrapper.state("username")).toEqual("test username");
    wrapper
      .find("#owisp-registration-email")
      .simulate("change", {target: {value: "test email", name: "email"}});
    expect(wrapper.state("email")).toEqual("test email");
    wrapper
      .find("#owisp-registration-password")
      .simulate("change", {target: {value: "testpassword", name: "password1"}});
    expect(wrapper.state("password1")).toEqual("testpassword");
    wrapper
      .find("#owisp-registration-password-confirm")
      .simulate("change", {target: {value: "testpassword", name: "password2"}});
    expect(wrapper.state("password2")).toEqual("testpassword");
  });

  it("should execute handleSubmit correctly when form is submitted", () => {
    axios
      .mockImplementationOnce(() => {
        return Promise.reject({
          response: {
            data: {
              username: "username error",
              email: "email error",
              detail: "nonField error",
              password1: "password1 error",
              password2: "password2 error",
            },
          },
        });
      })
      .mockImplementationOnce(() => {
        return Promise.resolve();
      });
    wrapper.setState({
      password1: "wrong password",
      password2: "wrong password1",
    });
    const event = {preventDefault: () => {}};
    wrapper.instance().handleSubmit(event);
    expect(
      wrapper.update().find(".owisp-registration-error-confirm"),
    ).toHaveLength(1);
    wrapper.setState({
      password1: "password",
      password2: "password",
    });
    return wrapper
      .instance()
      .handleSubmit(event)
      .then(() => {
        expect(wrapper.instance().state.errors).toEqual({
          username: "username error",
          email: "email error",
          nonField: "nonField error",
          password1: "password1 error",
          password2: "password2 error",
        });
        expect(wrapper.find(".owisp-registration-error")).toHaveLength(5);
      })
      .then(() => {
        return wrapper
          .instance()
          .handleSubmit(event)
          .then(() => {
            expect(wrapper.instance().state.errors).toEqual({});
            expect(wrapper.instance().state.success).toEqual(true);
            expect(
              wrapper.find(".owisp-registration-form.success"),
            ).toHaveLength(1);
            expect(
              wrapper.instance().props.authenticate.mock.calls.length,
            ).toBe(1);
          });
      });
  });
});
