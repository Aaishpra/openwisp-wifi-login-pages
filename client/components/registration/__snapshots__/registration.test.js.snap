// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> rendering should render correctly 1`] = `
<div
  className="owisp-registration-container"
>
  <form
    className="owisp-registration-form "
    onSubmit={[Function]}
  >
    <div
      className="owisp-registration-header"
    >
      <div
        className="owisp-registration-header-content"
      >
         sign up
      </div>
    </div>
    <div
      className="owisp-registration-fieldset"
    >
      <label
        className="owisp-registration-label owisp-registration-label-username"
        htmlFor="owisp-registration-username"
      >
        <div
          className="owisp-registration-label-text"
        >
          phone number or username
        </div>
        <input
          className="owisp-registration-input
                      owisp-registration-input-username
                      "
          id="owisp-registration-username"
          name="username"
          onChange={[Function]}
          pattern="[a-zA-Z@.+\\\\-_\\\\d]{1,150}"
          placeholder="enter phone number or username"
          required={true}
          title="only letters, numbers, and @/./+/-/_ characters"
          type="text"
          value=""
        />
      </label>
      <label
        className="owisp-registration-label owisp-registration-label-email"
        htmlFor="owisp-registration-email"
      >
        <div
          className="owisp-registration-label-text owisp-registration-label-text-email"
        >
          email
        </div>
        <input
          className="owisp-registration-input owisp-registration-input-email "
          id="owisp-registration-email"
          name="email"
          onChange={[Function]}
          placeholder="email address"
          required={true}
          title={null}
          type="email"
          value=""
        />
      </label>
      <label
        className="owisp-registration-label owisp-registration-label-password"
        htmlFor="owisp-registration-password"
      >
        <div
          className="owisp-registration-label-text"
        >
          password
        </div>
        <input
          className="owisp-registration-input owisp-registration-input-password
                      "
          id="owisp-registration-password"
          name="password1"
          onChange={[Function]}
          pattern=".{6,}"
          placeholder="password"
          required={true}
          title="password must be a minimum of 6 characters"
          type="password"
          value=""
        />
      </label>
      <label
        className="owisp-registration-label owisp-registration-label-confirm"
        htmlFor="owisp-registration-password-confirm"
      >
        <div
          className="owisp-registration-label-text"
        >
          confirm
        </div>
        <input
          className="owisp-registration-input owisp-registration-input-confirm "
          id="owisp-registration-password-confirm"
          name="password2"
          onChange={[Function]}
          placeholder="confirm password"
          required={true}
          title={null}
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      className="owisp-registration-add-info"
    >
      By signing up, you accept the 
      <a
        className="owisp-registration-additional-link"
        href="/default/registration/terms-and-conditions"
        onClick={[Function]}
      >
        terms and conditions
      </a>
       and the 
      <a
        className="owisp-registration-additional-link"
        href="/default/registration/privacy-policy"
        onClick={[Function]}
      >
        privacy policy
      </a>
       of this WiFi service. 
    </div>
    <input
      className="owisp-registration-form-btn owisp-registration-submit-btn"
      id="owisp-registration-submit-btn"
      type="submit"
      value="sign up"
    />
    <div
      className="owisp-login-links-div"
    >
      <a
        className="owisp-registration-link"
        href="/default/login"
        onClick={[Function]}
      >
        Already have an account? Log in
      </a>
    </div>
  </form>
</div>
`;
